lib/sgtn-client/api/translation.rb

```diff
@@ -1,158 +1,161 @@
-          obj = obj["data"]
-        end
-        return obj
+      items
+    end
+
+    def self.getTranslation(component, key, locale)
+      locale = SgtnClient::LocaleUtil.get_best_locale(locale)
+      items = get_cs(component, locale)
+      items&.dig('messages', key)
+    end
+
+    def self.get_cs(component, locale)
+      cache_key = SgtnClient::CacheUtil.get_cachekey(component, locale)
+      SgtnClient.logger.debug "[Translation][get_cs]cache_key=#{cache_key}"
+      expired, items = SgtnClient::CacheUtil.get_cache(cache_key)
+      if items.nil?
+        items = refresh_cache(component, locale).value # refresh synchronously if not in cache
+        # TODO: if an error occurs when requesting a bundle, need to avoid more requests
+      elsif expired && locale != LocaleUtil.get_source_locale # local source never expires.
+        refresh_cache(component, locale) # refresh in background
       end
 
+      items
+    end
+
+    def self.load(component, locale)
+      env = SgtnClient::Config.default_environment
+      mode = SgtnClient::Config.configurations[env]['bundle_mode']
+      SgtnClient.logger.debug "[Translation][load]mode=#{mode}"
+      if mode == 'offline'
+        load_o(component, locale)
+      else
+        load_s(component, locale)
+      end
+    end
+
+    def self.load_o(component, locale)
+      env = SgtnClient::Config.default_environment
+      product_name = SgtnClient::Config.configurations[env]['product_name']
+      version = SgtnClient::Config.configurations[env]['version'].to_s
+      translation_bundle = SgtnClient::Config.configurations[env]['translation_bundle']
+      bundlepath = translation_bundle + '/' + product_name + '/' + version + '/' + component + '/messages_' + locale + '.json'
+      SgtnClient::FileUtil.read_json(bundlepath)
+    end
+
+    def self.load_s(component, locale)
+      env = SgtnClient::Config.default_environment
+      product_name = SgtnClient::Config.configurations[env]['product_name']
+      vip_server = SgtnClient::Config.configurations[env]['vip_server']
+      version = SgtnClient::Config.configurations[env]['version'].to_s
+      url = vip_server + '/i18n/api/v2/translation/products/' + product_name + '/versions/' + version + '/locales/' + locale + '/components/' + component + '?checkTranslationStatus=false&machineTranslation=false&pseudo=false'
+      begin
+        obj = SgtnClient::Core::Request.get(url)
+      rescue StandardError => e
+        SgtnClient.logger.error e.message
+      end
+      obj && obj['data']
+    end
+
+    def self.fetch(component, locale)
+      # source locale always uses source bundles
+      return Source.getBundle(component) if LocaleUtil.is_source_locale(locale)
+
+      translation_bundle_thread = Thread.new { load(component, locale) }
+      source_bundle_thread = Thread.new { Source.getBundle(component) }
+      old_source_bundle = load(component, LocaleUtil.get_source_locale)
+      source_bundle = source_bundle_thread.value
```

Comment:
Not sure why get_cs was calling .value on current thread. 

I think this should now work.
```suggestion
      old_source_bundle = load(component, LocaleUtil.get_source_locale)
      source_bundle = get_cs(component, LocaleUtil.get_source_locale)
```

Reply:
Don't wok actually.